File: LeetCodeSolutions/LeetCode_0016.py
class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        n = len(nums)
        res = float('inf')
        for left in range(n - 2):
            middle, right = left + 1, n - 1
            if left > 0 and nums[left - 1] == nums[left]:
                continue
            while middle < right:
                curr_ans = nums[middle] + nums[left] + nums[right]
                curr_res = target - curr_ans
                if abs(res) >= abs(curr_res):
                    res = curr_res
                    ans = curr_ans
                if curr_res == 0:
                    return ans
                if curr_res > 0:
                    middle += 1
                if curr_res < 0:
                    right -= 1
            if nums[left] >= target:
                break
        return ans

File: LeetCodeSolutions/LeetCode_0016.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int ret = nums[0] + nums[1] + nums[2];
        int delta = abs(target - ret);
        for(int l = 0; l < n - 2; ++l){
            int m = l + 1, r = n - 1;
            if(l > 0 && nums[l - 1] == nums[l]) continue;
            if (delta == 0) break;
            while (m < r){
                int cur = nums[l] + nums[m] + nums[r];
                int cd = cur - target;
                int acd = abs(cd);
                if (acd < delta){
                    ret = cur;
                    delta = acd;
                }
                if (cd == 0){
                    ret = cur;
                    break;
                }
                else if(cd < 0){
                    while(m + 1 < r && nums[m] == nums[m + 1]) ++m;
                    ++m;
                }
                else if(cd > 0){
                    while (r - 1 > m && nums[r] == nums[r - 1]) --r;
                    --r;
                }

            }
        }
        return ret;
    }
};